kotlin/samples/tictactoe/common/src/main/java/com/squareup/sample/authworkflow/AuthWorkflow.kt

```diff
@@ -0,0 +1,140 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sample.authworkflow
+
+import com.squareup.sample.authworkflow.AuthService.AuthRequest
+import com.squareup.sample.authworkflow.AuthService.AuthResponse
+import com.squareup.sample.authworkflow.AuthService.SecondFactorRequest
+import com.squareup.sample.authworkflow.AuthState.Authorizing
+import com.squareup.sample.authworkflow.AuthState.AuthorizingSecondFactor
+import com.squareup.sample.authworkflow.AuthState.LoginPrompt
+import com.squareup.sample.authworkflow.AuthState.SecondFactorPrompt
+import com.squareup.sample.authworkflow.LoginScreen.SubmitLogin
+import com.squareup.sample.authworkflow.SecondFactorScreen.Event.CancelSecondFactor
+import com.squareup.sample.authworkflow.SecondFactorScreen.Event.SubmitSecondFactor
+import com.squareup.viewregistry.BackStackScreen
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.Workflow
+import com.squareup.workflow.WorkflowAction.Companion.emitOutput
+import com.squareup.workflow.WorkflowAction.Companion.enterState
+import com.squareup.workflow.WorkflowContext
+import com.squareup.workflow.rx2.onSuccess
+
+/**
+ * We define this otherwise redundant interface to keep composite reactors
+ * that build on [AuthWorkflow] decoupled from it, for ease of testing.
+ */
+interface AuthWorkflow : Workflow<Unit, AuthState, String, BackStackScreen<*>>
+
+/**
+ * Runs a set of login screens and pretends to produce an auth token,
+ * via a pretend [authService].
+ *
+ * Demonstrates both client side validation (email format, must include "@")
+ * and server side validation (password is "password").
+ *
+ * Includes a 2fa path for email addresses that include the string "2fa".
+ * Token is "1234".
+ */
+class RealAuthWorkflow(private val authService: AuthService) : AuthWorkflow {
+
+  override fun initialState(input: Unit): AuthState = LoginPrompt()
+
+  override fun compose(
+    input: Unit,
+    state: AuthState,
+    context: WorkflowContext<AuthState, String>
+  ): BackStackScreen<*> {
+    return when (state) {
+      is LoginPrompt -> {
+        BackStackScreen(LoginScreen(
+            state.errorMessage,
+            context.onEvent { event ->
+              when {
+                event.isValidLoginRequest -> enterState(Authorizing(event))
+                else -> enterState(LoginPrompt(event.userInputErrorMessage))
+              }
+            }
+        ))
+      }
+
+      is Authorizing -> {
+        context.onSuccess(
+            authService.login(AuthRequest(state.loginInfo.email, state.loginInfo.password))
+        ) { response ->
```

Comment:
Nit: Would recommend factoring these lambdas out into methods, both to keep `compose` tidy and for testability.

Reply:
I'm afraid disagree on both counts. I find it easier to keep the big picture of what's actually happening here by leaving it inline. 

I really don't understand the testability argument: this block should not be tested directly, the workflow should be tested as a black box: confirm that the mock service received a request; confirm expected rendering and output.

Reply:
#218 to follow up w/tests to prove my point.